feat(metadata): add SetAnnotation function

MartinBasti opened the pull request:
SetAnnotation sets one annotation into object (either creates new one or update existing one)

Commit:
feat(metadata): add SetAnnotation function

SetAnnotation sets one annotation into object (either creates new one or
update existing one)

Signed-off-by: Martin Basti <mbasti@redhat.com>

jinqi7 commented:
LGTM

davidmogar commented on metadata/metadata.go:
```diff
@@ -162,6 +162,23 @@ func HasLabelWithValue(obj v1.Object, key, value string) bool {
+
+	entries := obj.GetAnnotations()
+	entries[key] = value
+	obj.SetAnnotations(entries)
```
You can replace the 171 to 177 with:
```
        if annotations := obj.GetAnnotations(); annotations == nil {
		obj.SetAnnotations(map[string]string{key: value})
	} else {
		annotations[key] = value
	}
```

MartinBasti replied:
thanks!

davidmogar commented on metadata/metadata.go:
```diff
@@ -162,6 +162,23 @@ func HasLabelWithValue(obj v1.Object, key, value string) bool {
 }
 
+// SetAnnotation creates the annotation or updates the existing annotation.
+func SetAnnotation(obj v1.Object, key string, value string) error {
```
Could you add also a function for labels?

MartinBasti replied:
added in new commit

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,56 @@ var _ = Describe("Metadata", func() {
+	When("SetAnnotation is called", func() {
+		It("should add the annotation to the object", func() {
+			key := "foo"
+			value := "bar"
```
Other tests declare this as `annotations := map[string]string{"foo": "bar"}`. Could you do something similar. Then you could compare directly the maps. It applies to other tests.

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,56 @@ var _ = Describe("Metadata", func() {
+			Expect(pod.Annotations).To(HaveLen(1))
+			Expect(SetAnnotation(pod, key, value)).To(Succeed())
+			Expect(pod.Annotations).To(HaveLen(2))
+			Expect(pod.Annotations).To(HaveKeyWithValue(key, value))
```
Not checking here if the existing ones changed. Use `MatchAllKeys` (check other tests in this very same file).

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,56 @@ var _ = Describe("Metadata", func() {
+
+		})
+
+		It("should error if the object is nil", func() {
```
Could you move this to the beginning? It's nice when tests follow code order.

MartinBasti replied:
I copy pasted order of AddAnnotation tests, I can update it, no problem, JFYI

davidmogar replied:
Good catch. We can update that in a different PR ;)

Commit:
feat(metadata): add SetLabel function

SetLabel sets one label into object (either creates new one or
update existing one)

Signed-off-by: Martin Basti <mbasti@redhat.com>

davidmogar commented on metadata/metadata.go:
```diff
@@ -162,6 +163,36 @@ func HasLabelWithValue(obj v1.Object, key, value string) bool {
 	return ok && val == value
 }
 
+// SetAnnotation creates the annotation or updates the existing annotation.
```
What about this?
```suggestion
// SetAnnotation adds a new annotation to the referenced object or updates its value if it already exists.
```

davidmogar commented on metadata/metadata.go:
```diff
@@ -162,6 +163,36 @@ func HasLabelWithValue(obj v1.Object, key, value string) bool {
+	return nil
+}
+
+// SetLabel creates the label or updates the existing label.
```
What about this?
```suggestion
// SetLabel adds a new label to the referenced object or updates its value if it already exists.
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
 	})
 
+	When("SetAnnotation is called", func() {
+
```
```suggestion
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+			}
+			for key, value := range annotations {
+				Expect(pod.Annotations).To(HaveKeyWithValue(key, value))
+			}
```
What about setting more annotations and then compare the maps? Something like this:
```
                        annotations := map[string]string{"foo": "bar", "baz": "qux"}
			pod := &corev1.Pod{}

			Expect(pod.Annotations).To(HaveLen(0))

			for key, value := range annotations {
				Expect(SetAnnotation(pod, key, value)).To(Succeed())
			}

			Expect(pod.Annotations).To(Equal(annotations))
```

MartinBasti replied:
How exactly will this improve test coverage?

davidmogar replied:
In this case it will ensure that there are no extra annotations after running. But it also removes the second loop.

MartinBasti replied:
I used this with single annotation
```
Expect(pod.Annotations).To(Equal(annotations))
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+			Expect(pod.Annotations).To(HaveLen(1))
+			for key, value := range annotations {
+				Expect(pod.Annotations).To(HaveKeyWithValue(key, value))
+			}
```
If it has a single element at this point you could just check for the specific element without the loop:
```
Expect(pod.Annotations["foo"]).To(Equal("corge"))
```

MartinBasti replied:
I was trying to use annotations variable, I don't understand why I had to change 2 variables to map, when instead of using variables here, I'm supposed to hardcode it instead.

davidmogar replied:
I was just trying to remove the second loop, but fair point.

MartinBasti replied:
I removed second loop

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+	})
+
+	When("SetLabel is called", func() {
+
```
```suggestion
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+				Expect(pod.Annotations).To(HaveKeyWithValue(key, value))
+			}
+		})
+
```
```suggestion
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+				Expect(pod.Labels).To(HaveKeyWithValue(key, value))
+			}
+		})
+
```
```suggestion
```

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+
+	})
+
+	When("SetLabel is called", func() {
```
My previous comments apply to these tests here as they are mostly a 1:1 copy.

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+			Expect(pod.Annotations).To(HaveLen(1))
+			for key, value := range annotations {
+				Expect(SetAnnotation(pod, key, value)).To(Succeed())
+			}
```
Could you add empty lines before and after the loop. Looks cleaner.

davidmogar commented on metadata/metadata_test.go:
```diff
@@ -555,6 +555,126 @@ var _ = Describe("Metadata", func() {
+			}
+
+			Expect(pod.Annotations).To(HaveLen(1))
+			for key, value := range annotations {
```
Could you add empty lines before and after the loop. Looks cleaner.
